### PR TITLE
Use v2 api from distribution

### PR DESCRIPTION
**Pull request by dmcgowan**
Update registry package to use the v2 registry api from distribution. Update interfaces to directly take in digests.

### CODE REVIEW

**Comment by jfrazelle** [comment]
i think this should wait until the registry is refactored no? this is a huge patch

**Comment by dmcgowan** [comment]
The only non-vendor files touched is the registry/*.go and graph/*.go files, and the only non-refactor change is using the digest Parser instead of the string split.

File: graph/push.go [hunk 1]
```diff
@@ -378,13 +377,13 @@ func (s *TagStore) pushV2Repository(r *registry.Session, localRepo Repository, o
 
 			var exists bool
 			if len(checksum) > 0 {
-				sumParts := strings.SplitN(checksum, ":", 2)
```
Comment (stevvooe) [inline graph/push.go]: It's good to see these random string splits going away.

**Comment by stevvooe** [comment]
LGTM! Needs a rebase though.

**Comment by dmcgowan** [comment]
Rebased

> i think this should wait until the registry is refactored no? this is a huge patch

I do recall you requesting we break the distribution refactor into smaller PRs. This would be an example of such a PR, less than 50 lines of changes to graph/ and registry/ package, the remaining changes are related to moving and vendoring.

**Comment by stevvooe** [comment]
@jfrazelle This is just using the API descriptors and errors from the distribution project. The refactor is a different set of work, focused on using a provided client.

**Comment by jfrazelle** [comment]
oh sorry missed the whole vendor change thing going on here, carry on

**Comment by dmcgowan** [comment]
Carry on as move from design review to code review? :grinning:

**Comment by jfrazelle** [comment]
oh idk about that

**Comment by jfrazelle** [comment]
need another maintainer eyes to approve first :angel:

**Comment by dmp42** [comment]
ping @tiborvass @icecrime @LK4D4 @vbatts and the cat :)

**Comment by xiaods** [comment]
can't wait for this feature. +1

**Comment by icecrime** [comment]
Cool! LGTM

**Comment by vbatts** [comment]
LGTM

**Comment by tiborvass** [comment]
LGTM tested with private registry v1 and v2

**Pull request merged**